### ISSUE TITLE
dtest test_do_not_finish_view_building_with_hints fails with tablets

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#18427]
The dtest `materialized_views_test.py::TestMaterializedViews::test_do_not_finish_view_building_with_hint` fails with tablets enabled (`--tablets`). 

The test `test_do_not_finish_view_building_with_hints` does 

```
session.execute("CREATE MATERIALIZED VIEW t_by_v AS SELECT * FROM t " "WHERE v IS NOT NULL AND id IS NOT NULL PRIMARY KEY (v, id)")    
self.cluster.stop_nodes([node2, node3])
wait_for_view_build_start(session, "ks", "t_by_v")
```

But this wait fails, reporting: `View building didn't start in 20 seconds`.

**Comment by nyh** [comment]
Depending on timing (maybe this timing changed in the tablet case which is why we're only seeing this failure now?), the test kills two nodes 2,3 before view building on the surviving node 1 had a chance to start. 

The view building does begin, but the first step quickly fails because the remote nodes are down (and the view is not local - we have N=3, RF=1 and the view partition key is different) - we see in the log

```
INFO  2024-04-25 20:30:15,056 [shard 0:strm] view - Building view ks.t_by_v, starting at token minimum token
WARN  2024-04-25 20:30:15,057 [shard 0:strm] view - Error applying view update to 127.0.17.2 (view: ks.t_by_v, base token: -9223297786983086897, view token: -9223297786983086897): exceptions::unavailable_exception (Cannot achieve consistency level for cl ONE. Requires 1, alive 0)
```

Still, the test utility function `wait_for_view_build_start` which reads `views_builds_in_progress` from node 1 thinks the view build never started. Why? The `views_builds_in_progress` is actually a virtual reader over `scylla_views_builds_in_progress`. Is it possible that either the original table or the virtual table over it doesn't indicate the view building started before the first build step finished? This is just a guess but I don't yet have a better guess.

**Comment by bhalevy** [comment]
Seen again in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/618/testReport/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split005___test_do_not_finish_view_building_with_hints/

**Comment by nyh** [comment]
@bhalevy I'll look now at the MV implementation to see if I can prove my guess above.

**Comment by nyh** [comment]
The message `Building view ks.t_by_v, starting at token` comes from  `view_builder::add_new_view()`. Immediately after logging that message, this function writes into the distributed `system_distributed.view_build_status` (via start_view_build()) and into the local `system.scylla_views_builds_in_progress` (via system_keyspace::register_view_for_building()), and I don't see what can possibly cause the latter to not be done if the first build step fails. So I don't understand how the test's wait_for_view_build_start() can possibly not finished if the message "Building view ks.t_by_v" was printed.

I'm still struggling to understand this code. Quoting myself from earlier this year in yet another bug in this area: (https://github.com/scylladb/scylladb/issues/16646)

Unfortunately, I'm not familiar at all with this code - it was originally written by Duarte six years ago (commit https://github.com/scylladb/scylladb/commit/901faabaa2d4ceda0b41497a298717203d118894) and then a few years ago reorganized by @xemul. I will have to spend hours or days on it to get any understanding for the flow of this code. I recently created an issue https://github.com/scylladb/scylladb/issues/16590 about moving exactly this code into a separate source file (and issue #16569 about documenting this aspect of MV, and all other apects).

**Comment by nyh** [comment]
Ok, I think I finally got it.

I made progress understanding the test's utility function `wait_for_view_build_start()` (dtest's tools/ tables_view_manager.py), and I think it finally confirms my guess from https://github.com/scylladb/scylladb/issues/18427#issuecomment-2077833474 that this function is buggy.

The test function has:
```python
    def _check_build_started():
        result = rows_to_list(session.execute("SELECT last_token FROM system.views_builds_in_progress " f"WHERE keyspace_name='{ks}' AND view_name='{view}'"))
        return result != [[None]]
```

To consider the view build "started" it's not enough that the the row for the given view_name exists - it also needs to contain a non-null value for the `last_token` column. This column is build by the virtual reader (build_progress_virtual_reader.hh) from the scylla `next_token` column.  The function I mentioned above `system_keyspace::register_view_for_building()` which gets called sets only `first_token`, **not** `next_token`. The `next_token` is only set by `system_keyspace::update_view_build_progress()`, which is called after the first build step succeeded. So like I guessed, it seems that if the first build step never succeeds, the `next_token` column doesn't get written and the test thinks the view build hasn't started.

So, I think this is a test bug - a bug in `tools/tables_view_manager.py` `_check_build_started()`: This function should check just that the row appeared in `system.views_builds_in_progress`, or (but why bother?) check that `first_token` was set - not `next_token`.

Now I need to see if I can reproduce this test failure - and if I can do so reliably, I can check if my proposed patch to the test function fixes it.

**Comment by nyh** [comment]
Note that Cassandra's dtest also has `wait_for_view_build_start()` https://github.com/apache/cassandra-dtest/blob/trunk/materialized_views_test.py - but it just counts the rows - doesn't check for last_token.

So it seems **we** broke this function, but @fruch I can't figure out how because the pytest conversion wiped out all the history. Maybe you can help me figure out why/when we broke this `wait_for_view_build_start()` test function?

In any case, whether I understand why we broke it, now I know how to fix it :-)

**Comment by fruch** [comment]
@nyh 

this function introduced in our repo in:
https://github.com/scylladb/scylla-dtest/commit/66a40b70d15a2f4f865109dd7d6e0445f672c568

and in upstream in:
https://github.com/apache/cassandra-dtest/commit/59058a001c07974420c7299aa5741741f1bc6ba3

seems like it never copied the logic from upstream, so for the last 6 year this function exactly like it's now.

**Comment by nyh** [comment]
This issue was accidentally closed by a push to my private repository, so I reopened it.
I sent a dtest PR to fix this issue: https://github.com/scylladb/scylla-dtest/pull/4580 - it's a dtest bug, not a scylla core bug.

**Comment by kostja** [comment]
Seen again in https://jenkins.scylladb.com/job/scylla-6.0/job/dtest-debug/24/testReport/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split011___test_do_not_finish_view_building_with_hints/

**Comment by yarongilor** [comment]
reproduced in enterprise as well:
https://jenkins.scylladb.com/job/scylla-enterprise/job/dtest-release/443/testReport/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split006___test_do_not_finish_view_building_with_hints/

**Comment by Annamikhlin** [comment]
seen again on https://jenkins.scylladb.com/job/scylla-master/job/dtest-release/685/testReport/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split005___test_do_not_finish_view_building_with_hints/

**Comment by nyh** [comment]
> This issue was accidentally closed by a push to my private repository, so I reopened it. I sent a dtest PR to fix this issue: [scylladb/scylla-dtest#4580](https://github.com/scylladb/scylla-dtest/pull/4580) - it's a dtest bug, not a scylla core bug.

That pull request was incomplete so I closed it. @cezarmoise is now pushing a new and improved fix in https://github.com/scylladb/scylla-dtest/pull/4650.
